project2/kernel.c

```diff
@@ -268,6 +282,114 @@ void Kernel_Request_Terminate() {
     Dispatch();
 }
 
+void Kernel_Msg_Send() {
+    // Periodic tasks cannot send
+    if (Cp->priority == PERIODIC) {
+        OS_Abort(PERIODIC_MSG);
+        return;
+    }
+
+    // Check if process id is valid
+    if (!VALID_ID(request_info->msg_to)) {
+        OS_Abort(INVALID_REQ_INFO);
+        return;
+    }
+
+    PD *p_recv = &Process[request_info->msg_to];
+    MTYPE recv_mask = p_recv->req_params->msg_mask;
+
+    // Check if info.msg_to is waiting for a message of same type
+    if (p_recv->state == RECV_BLOCK && CMP_MASK(recv_mask, request_info->msg_mask)) {
+        // If yes, change state of waiting process to ready and sender to reply block
+        p_recv->state = READY;
+        Cp->state = REPLY_BLOCK;
+
+        // Add the message data and pid of sender to the receiving processes request info
+        p_recv->req_params->msg_ptr_data = request_info->msg_ptr_data;
+        p_recv->req_params->out_pid = Cp->process_id;
+    } else {
+        // If not, sender process goes to send block state
+        MSG *msg = &Messages[Cp->process_id];
+
+        // Save message
+        msg->data = request_info->msg_ptr_data;
+        msg->mask = request_info->msg_mask;
+        msg->receiver = request_info->msg_to;
+
+        Cp->state = SEND_BLOCK;
+    }
+}
+
+void Kernel_Msg_Recv() {
+    // Periodic tasks cannot receive
+    if (Cp->priority == PERIODIC) {
+        OS_Abort(PERIODIC_MSG);
+        return;
+    }
+
+    // Check if there is a message waiting for us
+    PID sender_pid;
+    MSG *msg = NULL;
+    for (sender_pid = 0; sender_pid < MAXTHREAD; sender_pid += 1) {
+        if (Messages[sender_pid].receiver == Cp->process_id) {
+            msg = &Messages[sender_pid];
+            break;
+        }
+    }
+
+    // Check if there is a message waiting for this process
+    if (msg != NULL && CMP_MASK(request_info->msg_mask, msg->mask)) {
+        // If yes, change state to ready and set msg data on Cp's request info
+        request_info->msg_ptr_data = msg->data;
+        request_info->out_pid = sender_pid;
+
+        Cp->state = READY;
+
+        // Sender process now waiting for reply
+        PD *sender = &Process[sender_pid];
+        sender->state = REPLY_BLOCK;
+
+        // Remove data from Messages
+        msg->data = NULL;
+        msg->receiver = MAXTHREAD;
```

Comment:
I get that `0` is a valid receiver, so that doesn't work, but maybe `-1` here instead of `MAXTHREAD`.
`-1` would be interpreted as max int since the `PID` type is unsigned

Reply:
ah, that is a better idea. I wasn't sure about making it `-1` because it was unsigned. If it just becomes max int than that is good.